gensim/similarities/termsim.py

```diff
@@ -207,24 +228,41 @@ def __init__(self, source, dictionary=None, tfidf=None, symmetric=True, positive
                 term_index for term_index, _
                 in sorted(
                     tfidf.idfs.items(),
-                    key=lambda x: (lambda term_index, term_idf: (term_idf, -term_index))(*x), reverse=True)]
-
-        column_nonzero = np.array([0] * matrix_order, dtype=_shortest_uint_dtype(nonzero_limit))
-        column_sum = np.zeros(matrix_order, dtype=dtype)
-        matrix = sparse.identity(matrix_order, dtype=dtype, format="dok")
-
-        for column_number, t1_index in enumerate(columns):
-            if column_number % self.PROGRESS_MESSAGE_PERIOD == 0:
-                logger.info(
-                    "PROGRESS: at %.02f%% columns (%d / %d, %.06f%% density, "
-                    "%.06f%% projected density)",
-                    100.0 * (column_number + 1) / matrix_order, column_number + 1, matrix_order,
-                    100.0 * matrix.getnnz() / matrix_order**2,
-                    100.0 * np.clip(
-                        (1.0 * (matrix.getnnz() - matrix_order) / matrix_order**2)
-                        * (1.0 * matrix_order / (column_number + 1))
-                        + (1.0 / matrix_order),  # add density correspoding to the main diagonal
-                        0.0, 1.0))
+                    key=lambda x: (lambda term_index, term_idf: (term_idf, -term_index))(*x),
+                    reverse=True,
+                )
+            ]
+
+        if dtype is np.float16 or dtype is np.float32:
+            similarity_type_code = 'f'
+        elif dtype is np.float64:
+            similarity_type_code = 'd'
+        else:
+            raise ValueError('Dtype %s is unsupported, use numpy.float16, float32, or float64.' % dtype)
+        nonzero_counter_dtype = _shortest_uint_dtype(nonzero_limit)
+
+        column_nonzero = np.array([0] * matrix_order, dtype=nonzero_counter_dtype)
+        if dominant:
+            column_sum = np.zeros(matrix_order, dtype=dtype)
+        if symmetric:
+            assigned_cells = set()
+        row_buffer = array('Q')
+        column_buffer = array('Q')
+        data_buffer = array(similarity_type_code)
```

Comment:
Here's an idea:

```python
# buffer population functions

def populate_buffer_symmetric(t1_index, t2_index, similarity):
    assigned_cells.add((t1_index, t2_index))
    column_buffer.append(t1_index)
    row_buffer.append(t2_index)
    data_buffer.append(similarity)

    assigned_cells.add((t2_index, t1_index))
    column_buffer.append(t2_index)
    row_buffer.append(t1_index)
    data_buffer.append(similarity)

def populate_buffer_assymetric(t1_index, t2_index, similarity):
    column_buffer.append(t1_index)
    row_buffer.append(t2_index)
    data_buffer.append(similarity)
```

Then you can do a little bit less work on those nested for-blocks.

Reply:
In fact, we can simplify this much further, see 0a7c6542.